packages/runtime/container-runtime/src/gc/gcSummaryStateTracker.ts

```diff
@@ -276,38 +290,38 @@ export class GCSummaryStateTracker {
 		// If the summary was not tracked by this client, the state should be updated from the downloaded snapshot.
 		const snapshotTree = result.snapshotTree;
 		const metadataBlobId = snapshotTree.blobs[metadataBlobName];
-		if (metadataBlobId) {
-			const metadata = await readAndParseBlob<IContainerRuntimeMetadata>(metadataBlobId);
-			this.latestSummaryGCVersion = getGCVersion(metadata);
-		}
+		const metadata = metadataBlobId
+			? await readAndParseBlob<IContainerRuntimeMetadata>(metadataBlobId)
+			: undefined;
+		this.latestSummaryGCVersion = getGCVersion(metadata);
 
 		const gcSnapshotTree = snapshotTree.trees[gcTreeKey];
 		// If GC ran in the container that generated this snapshot, it will have a GC tree.
 		this.wasGCRunInLatestSummary = gcSnapshotTree !== undefined;
-		let latestGCData: IGarbageCollectionSnapshotData | undefined;
-		if (gcSnapshotTree !== undefined) {
-			latestGCData = await getGCDataFromSnapshot(gcSnapshotTree, readAndParseBlob);
+
+		if (gcSnapshotTree === undefined) {
+			return undefined;
 		}
-		this.pendingSummaryData = undefined;
-		return latestGCData;
-	}
 
-	/**
-	 * Update state from the given snapshot data. This is done during load and during refreshing state from a snapshot.
-	 * @param gcSnapshotData - The GC snapshot data to update state from.
-	 */
-	public updateStateFromSnapshotData(gcSnapshotData: IGarbageCollectionSnapshotData | undefined) {
-		// If there is no snapshot data, it means this snapshot was generated with GC disabled. Unset all GC state.
-		if (gcSnapshotData === undefined) {
-			this.latestSummaryData = undefined;
-			return;
+		let snapshotData = await getGCDataFromSnapshot(gcSnapshotTree, readAndParseBlob);
+
+		// If the GC version in the snapshot does not match the GC version currently in effect, the GC data
+		// in the snapshot cannot be interpreted correctly. Set everything to undefined except for deletedNodes
+		// because irrespective of GC versions, these nodes have been deleted and cannot be brought back. The
+		// deletedNodes info is needed to identify when these nodes are used.
+		if (getGCVersion(metadata) !== this.currentGCVersion) {
+			snapshotData = {
```

Comment:
@markfields I was thinking that if refreshing from snapshot, we don't need to update deleted nodes right. We need to do it when loading from a base snapshot but when refreshing, we already have deleted nodes from whatever the last GC run was and we can continue using that.
The only problem with updating it here is that if we move back in time, the deleted nodes unncessarily move back in time too.

Reply:
As discussed elsewhere, makes sense to me, since it's not used for incremental summary calculation. We can keep that in mind when we revisit the state machine and different rolls the different in-memory things have.